### PR TITLE
Add a doxygen module on the interplay between FEValues, FE and Mapping.

### LINKED ISSUE #1308: FEValues cleanup: Write a module about the interaction between Mapping, FE, FEValues

### PR DESCRIPTION
**Pull request by bangerth**
This is a lot of documentation, but I feel that it has been missing for a long
time and having worked through this recently, have made sure it is up to
date. (In fact, it is too up to date: It references FiniteElement::requires_update_flags()
which does not even exist yet, but will soon.)

Fixes #1308. In reference to #1198.

### CODE REVIEW

File: doc/doxygen/headers/fe_vs_mapping_vs_fevalues.h [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+ * reference cell, and we need to map these gradients to the real cell $K$.
+ * Similar operations can also be defined for matrices (tensors of rank 2,
+ * as opposed to vectors which are tensors of rank 1) and higher order tensors.
+ *
+ * Many of these mappings do not only need the map $\mathbf F_K$ itself,
+ * but also the gradients of this mapping, typically referred to as the
+ * Jacobian $J_K=\hat\nabla \mathbf F_K$, as well as higher derivatives.
+ *
+ * Since FEValues only ever needs to evaluate these things at quadrature
+ * points, mappings do not in general need to provide the ability to
```
Comment (kronbichler) [inline doc/doxygen/headers/fe_vs_mapping_vs_fevalues.h]: provide the ability to _evaluate_ ?

File: doc/doxygen/headers/fe_vs_mapping_vs_fevalues.h [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+ * This already happens at the time the FEValues constructor is run.
+ * Because the mapping does not depend on the finite element (though the
+ * latter does depend on the former), FEValues first asks the finite
+ * element via FiniteElement::requires_update_flags() which <i>other</i>
+ * pieces of information it also requires to make the user request
+ * happen. As an example, if the finite element were of type
+ * FE_Q, then it would determine that in order to compute the
+ * gradients of the shape functions on the real cell $K$, it will
+ * need to compute the gradients of the shape functions on the
+ * reference cell (something it can affect itself, without any
```
Comment (kronbichler) [inline doc/doxygen/headers/fe_vs_mapping_vs_fevalues.h]: 'affect itself' sounds strange (but I'm not a native English speaker). Maybe 'do on its own' instead?

File: doc/doxygen/headers/fe_vs_mapping_vs_fevalues.h [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+ *
+ * <h2>Computing on a given cell</h2>
+ *
+ * All of the previous steps happened at the time the FEValues object
+ * was created. Up to this point, all we did was set up data structures,
+ * but nothing useful has been computed so far from the perspective of
+ * the user. This only happens when FEValues::reinit() is called on
+ * a concrete cell $K$.
+ *
+ * The thing FEValues then does are, in this order:
```
Comment (kronbichler) [inline doc/doxygen/headers/fe_vs_mapping_vs_fevalues.h]: Please re-visit this sentence. 'the thing' and 'are' seem to not go together.

**Comment by kronbichler** [comment]
Apart from a few places where I would suggest re-phrasing, I'm fine with this description. If there are no further comments by someone else today, merge yourself.

File: doc/doxygen/headers/fe_vs_mapping_vs_fevalues.h [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+ * and so FE_Q's implementation of FiniteElement::requires_update_flags()
+ * function (provided in the intermediate class FE_Poly) will return
+ * both the original update_gradients flag as well as
+ * update_covariant_transformation.
+ *
+ * In a second step, the FEValues object will then call the corresponding
+ * function in the mapping, Mapping::requires_update_flags() to determine
+ * what is required to provide both update_gradients and
+ * update_covariant_transformation. The former is not within the realm
+ * of the mapping, so is ignored. The latter will of typically require
```
Comment (agrayver) [inline doc/doxygen/headers/fe_vs_mapping_vs_fevalues.h]: 'will of typically' -- remove of?

File: doc/doxygen/headers/fe_vs_mapping_vs_fevalues.h [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+ *   Mapping::get_subface_data()), again with the quadrature object and
+ *   the final set of update flags. These functions do essentially the
+ *   same as their counterparts in the mappings, and again the object
+ *   so initialized, this time of a type derived from
+ *   FiniteElement::InternalDataBase, will always be given back to the finite
+ *   element whenever the FEValues object wants something from the finite
+ *   element object at a later time.
+ *
+ * This approach allows us to use finite element and mapping objects from
+ * multiple FEValues objects at the same time, and possible from multiple
```
Comment (agrayver) [inline doc/doxygen/headers/fe_vs_mapping_vs_fevalues.h]: possibly?

**Comment by davydden** [comment]
i wonder if there is already a similar detailed description on parts of FE which are related to hp::DoFHandler? Maybe that could be mentioned here as it's related to writing a FE although being a more advanced topic.

**Comment by bangerth** [comment]
@davydden : I think this is the wrong place, but I've got something like this on my list -- #1397

**Comment by davydden** [comment]
@bangerth ok

**Comment by bangerth** [comment]
Having heard no other complaints, I'm following @kronbichler 's suggestion and merge.

Thanks, Martin and @agrayver for the input!

**Pull request merged**